python/lsst/daf/butler/_registry_shim.py

```diff
@@ -0,0 +1,394 @@
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = ("Registry",)
+
+import contextlib
+from collections.abc import Iterable, Iterator, Mapping, Sequence
+from typing import TYPE_CHECKING, Any
+
+from .core import (
+    DataCoordinate,
+    DataId,
+    DatasetAssociation,
+    DatasetId,
+    DatasetIdGenEnum,
+    DatasetRef,
+    DatasetType,
+    Dimension,
+    DimensionElement,
+    DimensionGraph,
+    DimensionRecord,
+    DimensionUniverse,
+    NameLookupMapping,
+    Timespan,
+)
+from .registry import Registry
+from .registry._collection_summary import CollectionSummary
+from .registry._collectionType import CollectionType
+from .registry._defaults import RegistryDefaults
+from .registry.queries import DataCoordinateQueryResults, DatasetQueryResults, DimensionRecordQueryResults
+
+if TYPE_CHECKING:
+    from ._butler import Butler
+    from .registry._registry import CollectionArgType
+    from .registry.interfaces import ObsCoreTableManager
+
+
+class RegistryShim(Registry):
+    """Implementation of `Registry` interface exposed to clients by `Butler`.
+
+    Parameters
+    ----------
+    butler : `Butler`
+        Data butler instance.
+
+    Notes
+    -----
+    This shim implementation of `Registry` forwards all methods to an actual
+    Registry instance which is internal to Butler or to Butler methods. Its
+    purpose is to provide a stable interface to many client-visible operations
+    while we perform re-structuring of Registry and Butler implementations.
+    """
+
+    def __init__(self, butler: Butler):
+        self._butler = butler
+        self._registry = butler._registry
+
+    def isWriteable(self) -> bool:
+        # Docstring inherited from a base class.
+        return self._registry.isWriteable()
+
+    @property
+    def dimensions(self) -> DimensionUniverse:
+        # Docstring inherited from a base class.
+        return self._registry.dimensions
+
+    @property
+    def defaults(self) -> RegistryDefaults:
+        # Docstring inherited from a base class.
+
+        # Note that we do not allow changing defaults via this class
```

Comment:
Looks like we do, given the setter below.

Reply:
Indeed, I wanted to disable setter but then had to re-enable it because there are some legit clients, but forgot to remove the comment, will drop it.